### PR TITLE
cpu/samd21: implement low power modes

### PR DESCRIPTION
**Pull request by LudwigKnuepfer**
TODO: needs testing

### CODE REVIEW

File: cpu/samd21/lpm_arch.c [hunk 1]
```diff
@@ -14,40 +15,111 @@
...
+            start_lpm();
+            break;
+        case LPM_POWERDOWN:             /* Sleep mode Idle 2 */
+            /* Idle Mode 2 */
+            SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+            PM->SLEEP.reg = SYSTEM_SLEEPMODE_IDLE_2;
+            start_lpm();
+            break;
+        case LPM_OFF:                   /* Standby Mode - Potential Wake Up sources: Asynchronous */
+            current_mode = LPM_POWERDOWN;
```
Comment (saurabh984) [inline cpu/samd21/lpm_arch.c]: My mistake sorry! Should be
```
current_mode = LPM_OFF;
```

File: cpu/samd21/lpm_arch.c [hunk 2]
```diff
@@ -14,40 +15,111 @@
...
+            break;
+        case LPM_SLEEP:                 /* Sleep mode Idle 1 */
+            current_mode = LPM_SLEEP;
+             /* Idle Mode 1 */
+            SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+            PM->SLEEP.reg = SYSTEM_SLEEPMODE_IDLE_1;
+            start_lpm();
+            break;
+        case LPM_POWERDOWN:             /* Sleep mode Idle 2 */
+            /* Idle Mode 2 */
```
Comment (saurabh984) [inline cpu/samd21/lpm_arch.c]: My mistake sorry! Missing line
```
current_mode = LPM_POWERDOWN;
```

File: cpu/samd21/lpm_arch.c [hunk 3]
```diff
@@ -14,40 +15,111 @@
...
 {
-    // TODO
+    return current_mode;
 }
 
-void lpm_arch_end_awake(void)
+void lpm_arch_awake(void)
 {
-    // TODO
+    if (current_mode == LPM_SLEEP) {
```
Comment (saurabh984) [inline cpu/samd21/lpm_arch.c]: Not quite sure which modes will require a ```cpu_init();``` need to test this!

**Comment by LudwigOrtmann** [comment]
Just tried with the diff below and couldn't see any difference on our energy consumption test equipment.
We tried to measure the energy consumption over the external power pins which might be imprecise. The xpro has support for energy measurements on the jumper'd pins next to the buttons.. we didn't try that.
Maybe someone else can try this PR with their measurement tools?

```diff
diff --git a/core/kernel_init.c b/core/kernel_init.c
index a57f119..c18b6e6 100644
--- a/core/kernel_init.c
+++ b/core/kernel_init.c
@@ -64,8 +64,8 @@ static void *idle_thread(void *arg)
             lpm_set(LPM_IDLE);
         }
         else {
-            lpm_set(LPM_IDLE);
-            /* lpm_set(LPM_SLEEP); */
+            //lpm_set(LPM_IDLE);
+            lpm_set(LPM_SLEEP);
             /* lpm_set(LPM_POWERDOWN); */
         }
```

**Comment by saurabh984** [comment]
I'll have access to my lab again next week. I'll give it a go then. Thanks

File: cpu/samd21/lpm_arch.c (hunk 2 above)
Comment (LudwigOrtmann) [inline cpu/samd21/lpm_arch.c]: did

**Comment by LudwigOrtmann** [comment]
@saurabh984 I addressed your comments.

File: cpu/samd21/lpm_arch.c [hunk 4]
```diff
@@ -14,40 +15,112 @@
  * @brief       Implementation of the kernels power management interface
  *
  * @author      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author      Saurabh Singh <email-missing@example.com>
```
Comment (LudwigOrtmann) [inline cpu/samd21/lpm_arch.c]: @saurabh984 your email is missing here ;)

Comment (saurabh984) [inline cpu/samd21/lpm_arch.c]: @LudwigOrtmann saurabh@cezy.co Haven't had a chance to try it yet :(

**Comment by PeterKietzmann** [comment]
@LudwigOrtmann I tried to test this PR with:
```
static void button_handler(void *args)
{
    puts("Button Handler");
    lpm_arch_awake();
}

int main(void)
{
    puts("Test LPM modes\n");
    /* this is the user button on samr21-xpro */
    gpio_init_int(GPIO_1, GPIO_NOPULL, GPIO_BOTH, &button_handler, 0);

    puts("Init lpm");
    lpm_arch_init();

    puts("LPM_ON");
    lpm_arch_set(LPM_ON);

    puts("LPM_IDLE");
    lpm_arch_set(LPM_IDLE);

    puts("LPM_SLEEP");
    lpm_arch_set(LPM_SLEEP);

    puts("LPM_POWERDOWN");
    lpm_arch_set(LPM_POWERDOWN);

    puts("LPM_OFF");
    lpm_arch_set(LPM_OFF);
}
```
I couldn't awake the board after `LPM_SLEEP`. Can you tell how this is done?

**Comment by saurabh984** [comment]
@PeterKietzmann Try commenting out the cpu_init() (shown below) code from ```lpm_arch_awake(void)``` please?

```
if (current_mode == LPM_SLEEP) {
        /* Re-init */
        cpu_init();
    }
```

**Comment by PeterKietzmann** [comment]
@saurabh984, I commented this line out. With this change I could also awake from `LPM_SLEEP` and enter `LPM_POWERDOWN`. But I did not came out of `LPM_POWERDOWN` (btw: I haven't looked into the boards documentation). Then I measured the current through the "current measurement header". My results:
`LPM_ON/LPM_IDLE` : 560 uA
`LPM_SLEEP` : 535 uA
`LPM_POWERDOWN` : 500 uA
I don't know if this are the expected values: In addition: First I tried to measure the current through a 10Ohm shunt-resistor and an Oscilloskop which "crashed" the board in any way. Then I took my multimeter (ammeter) and had the same problems. Once I switched the multimeter-range to MICROampere the internal resistance seemd to be small enough to fit the measurement.

**Comment by PeterKietzmann** [comment]
@saurabh984 did you experience the same?

**Comment by bapclenet** [comment]
@PeterKietzmann, from p1048 to 1051/1138 (samr21 datasheet), you can find the power consumption of the samr21. There is also a schematic about how to use the measurement pin. 
Will try it in the comings days hopefully

**Comment by bapclenet** [comment]
Here are my results with @PeterKietzmann source code (above):
<code>LPM_IDLE/LPM_ON</code>: 1,62mA
<code>LPM_SLEEP</code>1,38mA
<code>LPM_POWERDOWN</code> : 1,08mA
<code>LPM_OFF</code>: 0,39mA

Can't wake up from LPM_OFF

By the way, @PeterKietzmann, you should use a <code>GPIO_PULLUP</code> and <code>GPIO_RISING</code> otherwise the interrupt doesn't work properly.
 <code>gpio_init_int(GPIO_1, GPIO_PULLUP, GPIO_RISING, &button_handler, 0);</code>

**Comment by bapclenet** [comment]
@saurabh984 , may I ask why do you usethe System Control Block (SCB). It's not mention for power managing. Which documentation is it related to?

**Comment by gebart** [comment]
@bapclenet what do you mean by the SCB is not mentioned for power management?
The SLEEPDEEP bit in the SCB SCR decides if the CPU should go to STOP mode or WAIT mode when `wfi` is executed.

**Comment by bapclenet** [comment]
Sorry, my bad, I didn't know that before, I've just seen it.

**Comment by PeterKietzmann** [comment]
@bapclenet your results seem somehow reasonable even if they appear to be kind of small, comparing to the datasheet (worse are mine :-) ). But honestly I didn't compare the clock setup etc. . Now the question is: is there anybody out who has a deeper knowledge with this MCU and with (low) power modes? Otherwise I need to spend some time in going into detail.

**Comment by PeterKietzmann** [comment]
@bapclenet can you explain in detail how you build your setup? Looking at the [datasheet](http://www.atmel.com/Images/Atmel-42223-SAM-R21_Datasheet.pdf) on page 1054 it feels like I can't  directly access the needed pins or have to use an external power supply __not__ from USB.
In the [user guide](http://www.atmel.com/Images/Atmel-42243-SAMR21-Xplained-Pro_User-Guide.pdf) on page 7 sec. 3.3.1 you can read that it's possible to measure the consumption through the pins on which the jumper is on. They have a different naming scheme to the pins mentioned in the datasheet. What are your doubts using these?

Sorry for all these questions. I don't want to spend too much time in this and you already did these steps.

**Comment by bapclenet** [comment]
@PeterKietzmann, I use a Multimeter from Agilent. 
+ in VCC target 
Ground in VCC MCU

There are the same pins as in the datasheet. What I'm doing is removing the header ([Datasheet](http://www.atmel.com/Images/Atmel-42223-SAM-R21_Datasheet.pdf) p4 - CURRENT MEASUREMENT HEADER) and plug them to your multimeter.

I think (not sure about that), whatever the source is, there is a regulator which transforms 5V to 3.3V, and after the transformation, there are those two pins where you can measure the current before going to the MCU.

Does it make sens?

**Comment by bapclenet** [comment]
For more details:

**Comment by PeterKietzmann** [comment]
Yes and no. Measuring the current through the "current measurement header"-pins is reasonable and the way I already did (some time ago). What I don't understand is the hint you gave me with the schematic in the datasheet (p. 1054). Anyway, if I see it right and understand correctly you just power your board by the pins and not by an USB connector and you measure the current through the measurement extension header.

**Comment by PeterKietzmann** [comment]
Ok, at least I measured the same as you now :-) . Will look in the registers and the lpm idea a bit deeper in the next week and hopefully ACK this PR soon. Thanks for your support!

**Comment by bapclenet** [comment]
You mean p1051:
image

I think it is the same, you've got the battery below AMP 0 (which means VCC_TARGET on the Explained Board) and the rest (which is VCC_MCU on the Explained Board)


Powering by the pin seems better to me since USB doesn't consume anything then. But I would add that:
- Measuring with usb (no debbuging) doesn't seem to consume more.
- While debugging, the value change as well as the behavior (I can go to <code>LPM_OFF</code> and interrupt is still working)

Let me know about your deeper understanding :-)

**Comment by PeterKietzmann** [comment]
@bapclenet I am fine with the code except one thing that I feel unsure about, the `cpu_init()`. Do you know why exactly this is placed to wakeup to board from `SYSTEM_SLEEPMODE_IDLE_1` mode? Also it appeard that I needed to comment this line out to awake the board after the `LPM_SLEEP` command. Any knowledge on this? Compare also the [datasheet](http://www.atmel.com/Images/Atmel-42223-SAM-R21_Datasheet.pdf) p. 119, I assume there is a relation.

**Comment by bapclenet** [comment]
@PeterKietzmann, no I don't know yet.
[Datasheet](http://www.atmel.com/Images/Atmel-42223-SAM-R21_Datasheet.pdf) p120: "The CPU and affected modules are restarted."
Does it mean that it starts with the previous state or it is reseted and we have to init the CPU?

Here are the issues we have to deal with:
 - <code>cpu_init()</code> needs some test to know what it is useful for. Why do we need it to wake up from LPM_SLEEP?@saurabh984, could you help us on this?
 - How to wake up from <code>STANDBY</code>? I couldn't when I tried.  
 - What about <code>SleepWalking</code>? How does it work? And could we implement it in the same time?
 - If I'm not wrong, each peripheral is able to run in <code>STANDBY</code> mode if its <code>RUNSTDBY</code> bit is enabled? Do we do something with that?

**Comment by saurabh984** [comment]
@bapclenet ```cpu_init()``` shouldn't be needed in any of the modes from what I understand. Both ```LPM_POWERDOWN``` and ```LPM_OFF``` require  
> Potential Wake Up sources: Asynchronous

as per the documentation. Sorry I can't really help here :(

As for ```SleepWalking``` and ```RUNSTDBY```, I'm not sure the API, as is, can support that.

**Comment by PeterKietzmann** [comment]
@bapclenet I think it is as @saurabh984 said. Comparing p. 119 table 14-3 you can see that `IDLE 0` and `IDLE 1` can be woke up by either synchronous or asynchronous sources. I assume we awoke from these states by the synchronous interrupt on the APB clock. States `IDLE 2` and `STANDBY` you need to wake-up asynchronously (_Note: interrupt generated on generic clock or external event_). That is the thing that's missing here and the reason why we can't awake from `LPM_POWERDOWN` mode in RIOT (which leads to `IDLE 2`). 
You seem to be experienced with this controller. Would you mind to take this PR over? Do you have the time? I rather don't have currently..

**Comment by PeterKietzmann** [comment]
@bapclenet may I assign you? Would you take on the work or don't you have time for this?

**Comment by bapclenet** [comment]
Ok @PeterKietzmann. 
But stay around if I need details or some testing on another board than mine. If you find someone interested in it or who could have more knowledge than us, invite him to help us :-)

**Comment by PeterKietzmann** [comment]
Thanks! Yes of course I will follow this PR. Ideally it is just the wakup from IDLE2 and STANDBY to realize. The rest works well. Also I won't try to implement all possible low power modes as our general API doesn't provide this.

**Comment by bapclenet** [comment]
It's not that we don't wake up from ``IDLE2``, it seems that UART is locked.
In uart.c, program is locked at ``while (UART_0_DEV.INTFLAG.bit.DRE == 0);``
Just by removing the puts in the main from your example @PeterKietzmann , it works

**Comment by PeterKietzmann** [comment]
But isn't it weird to debug in a low power mode?

**Comment by bapclenet** [comment]
Yes I think this is why the behaviour is unexpected :-)

**Comment by PeterKietzmann** [comment]
Dammit :-) ! Do you agree that we miss an "asynchronous wake-up source"?  Maybe we could open an issue then. Perhaps someone has a deeper knowledge in this CPU and knows about the wake-up sources...

**Comment by bapclenet** [comment]
While pressing SW0, it is an asynchronous wake-up definitely, it doesn't depend on any internal clock.
Yeah, I would be happy with someone with a better knowledge on this. Let's ask on ATMEL forum as well! This feature is mandatory for the SAMR21!

**Comment by bapclenet** [comment]
Some bugs might need to be taken into account ...
1:
>In Standby, Idle1 and Idle2 sleep modes the device might not wake up
from sleep. An External Reset, Power on Reset or Watch Dog Reset will start
the device again. Errata reference: 13140
Fix/Workaround:
the SLEEPPRM bits in the NVMCTRL.CTRLB register must be written to 3
(NVMCTRL - CTRLB.bit.SLEEPPRM = 3) to ensure correct operation of the
device. The average power consumption of the device will increase with 20uA
compared to numbers in the electrical characteristics chapter.

2:
>Digital pin outputs from Timer/Counters, AC (Analog Comparator),
GCLK (Generic Clock Controller), and SERCOM (I2C and SPI) do not change
value during standby sleep mode. Errata reference: 12537
Fix/Workaround:
Set the voltage regulator in Normal mode before entering STANDBY sleep mode
in order to keep digital pin output enabled. This is done by setting the RUNSTDBY
bit in the VREG register.

**Comment by bapclenet** [comment]
Moreover, from EIC:
>All interrupts are available in all sleep modes, but the EIC can be configured to automatically mask some interrupts in order to prevent device wake-up.
The EIC will continue to operate in any sleep mode where the selected source clock is running. The EIC’s interrupts can be used to wake up the device from sleep modes.

Which means that GCLK for EIC has to run

**Comment by bapclenet** [comment]
In ``cpu.c``, enable ``RUNSTBY`` for OSC8M like this:
    ``SYSCTRL->OSC8M.bit.RUNSTDBY = 1;``
And you will be able to wake up from STANDBY sleep
It means that OSC8M is running, I haven't measured the consumption. Let's see how we can do then because leaving OSC8M in STANDBY sleep might not be the perfect solution

**Comment by daniel-k** [comment]
If you need further help debugging this just ping me. I think I have a quite good understanding of this platform now after implementing duty-cycling to `STANDBY` in #3730 and would be happy to see this merged.

**Comment by bapclenet** [comment]
@daniel-k, Feel free to continue this PR, I don't have time to work on it. If you do, explain your thought while developping it.

**Comment by bapclenet** [comment]
@daniel-k, any work on it?

**Comment by daniel-k** [comment]
@bapclenet Not yet ;) There's not so much to here as the code work's as it is IIRC, but since the LPM architecture is supposed to get a general overhaul, I was more or less waiting for the release to finished.

**Comment by immesys** [comment]
I have a pretty nice jig for accurate average power measurement across a wide dynamic range (uA accuracy even when the target spikes to tens of mA periodically). If it can be of any use here, please let me know. I have both a samr21xpro and a board with an isolated samr21 sans the EDBG chip and other peripherals that mess with power measurement.

**Comment by bapclenet** [comment]
@immesys, yes it can be useful. Could you try this PR and tell use the results you get for each mode?

Try it with this simple code:
```
static void button_handler(void *args)
{
    puts("Button Handler");
    lpm_arch_awake();
}

int main(void)
{
    puts("Test LPM modes\n");
    /* this is the user button on samr21-xpro */
    gpio_init_int(GPIO_1, GPIO_NOPULL, GPIO_BOTH, &button_handler, 0);

    puts("Init lpm");
    lpm_arch_init();

    puts("LPM_ON");
    lpm_arch_set(LPM_ON);

    puts("LPM_IDLE");
    lpm_arch_set(LPM_IDLE);

    puts("LPM_SLEEP");
    lpm_arch_set(LPM_SLEEP);

    puts("LPM_POWERDOWN");
    lpm_arch_set(LPM_POWERDOWN);

    puts("LPM_OFF");
    lpm_arch_set(LPM_OFF);
}
```

**Comment by immesys** [comment]
I have not forgotten about this, I have spent the past few days designing an automated generic solution for power consumption CI testing. After all, I suspect we will want to check power consumption quite a bit as the LPM infrastructure evolves, and https://xkcd.com/1205/

**Comment by PeterKietzmann** [comment]
@immesys thanks for your efforts, sounds great! It would be really interesting to see your results and some insights to your setup once your're ready :-)

**Comment by OlegHahm** [comment]
@immesys, maybe you want to contact the shepherds of the [APMTF](https://github.com/RIOT-OS/RIOT/wiki/Active-Power-Management-Task-Force) about this.

**Comment by immesys** [comment]
Quick question. If rebased onto master, you wouldn't expect this code to stop working right? I've been working more on the power measurement infrastructure than actually verifying we are entering low power modes, but preliminary measurements with a program like:
```
#include <stdio.h>
#include <arch/lpm_arch.h>

int main(void)
{
    lpm_arch_init();
    while(1) lpm_arch_set(LPM_POWERDOWN);
    return 0;
}
```

don't show significant reductions in power consumption. Could it be that the radio or something is still on? I had to pull in the changes made in master because the actual DUT requires some of that code.

**Comment by immesys** [comment]
I should add that the platform support pulls in xtimer. Perhaps xtimer starts a peripheral that does not get stopped by the low power modes?

**Comment by immesys** [comment]
Belay that. I found the culprit. I'll have real answers imminently.

**Comment by daniel-k** [comment]
I  had to reset the MCU for low power modes to take effect although I have no idea why this was needed. 

And yes, you need to put the radio to sleep manually. I reached something around 20uA or so in LPM_OFF and some mA in LPM_POWERDOWN mode.

**Comment by immesys** [comment]
Ok, so my automated power measurement stuff is working. The TL;DR is 
 - you make a pull request on https://github.com/PowerProfiler/RIOT
 - a bot flashes it to a board and tests it against an array of apps that are on https://github.com/PowerProfiler/TestApplications
 - after testing for a few hours, the bot comments on the pull request with the results

The full readme is in the TestApplications repo linked above.

I have tested it against this pull request, and the applications I wrote are not getting great results, mostly near 1.6 - 1.9 mA. You can see the report [here](https://github.com/PowerProfiler/RIOT/pull/2). I would appreciate pull requests on the applications if there is a mistake in my logic.

The actual bot that runs the tests is also brand new code that I wrote over the past three days, so there is a lot of room for improvement there too.

**Comment by PeterKietzmann** [comment]
@immesys thanks a lot. Your PowerProfiler looks great at first sight. Unfortunately I currently don't find time to test and review the application, but hopefully will find some time soon. IMO we can merge this PR once the power modes are confirmed **but** switch to the  [APMTF](https://github.com/RIOT-OS/RIOT/wiki/Active-Power-Management-Task-Force)  then

**Comment by OlegHahm** [comment]
Doesn't seem to be realistic for the upcoming release next week.

**Comment by kYc0o** [comment]
I think this PR will need a lot of attention so I don't think it will make it for the release. What's the status BTW?

**Comment by PeterKietzmann** [comment]
@MichelRottleuthner did you already give it a try?

**Comment by MichelRottleuthner** [comment]
@PeterKietzmann Nope. It's on my todo list but I think it will take a few weeks before I have time for that.

**Comment by kYc0o** [comment]
Maybe @aabadie wants to take a look ;)

**Comment by immesys** [comment]
Hey all. Can I help get this merged in for the upcoming release? I have used it, along with a few other changes, and have a SAMR21 running at 5.2uA idle and 15uA average sampling temperature at 1Hz

**Comment by biboc** [comment]
@immesys, you're welcome to help us merging this PR :-)
How did you measure the current (can you provide a schema) ?
Did you try all different modes?
I remember that we had some problem to wake up from LPM_OFF, have you tried?
If you don't have a lot of changes, you can create a PR on LudwigKnuepfer:samd21_lpm branch (with rebase) or if it is totally different, create a new one.

**Comment by immesys** [comment]
So, the good news is yes, it is waking up fine from lpm off, the bad news is that the code is a little messy (as things get when you are just trying to make it work). I am busy splitting things up into independent PRs, so things like adding radio sleep to netdev, switching the xtimer source to ULP OSC32. 

I could really use a shepherd, as some of these changes are in core code and change default behavior (I use FLL instead of PLL saving a bit of energy, and I source it from the ULP instead of the 8mhz RC, which costs accuracy but saves energy). I am busy running experiments to quantify the benefit of each change individually so that the riot core devs can decide if each one is worth it.

**Comment by OlegHahm** [comment]
@kaspar030, @gebart, would you be willing to act as a shepherd? I would also volunteer but have only time in about two weeks from now.

**Comment by PeterKietzmann** [comment]
@immesys it's great to see your efforts improving the sam* support in RIOT! Would it help you to merge this PR "as is" (or with slight adoptions which you would need to propose ;-) )?

**Comment by immesys** [comment]
I use this file as-is, so it would definitely help to merge it. The slight changes elsewhere are too invasive to just merge, I will need to discuss them with the core maintainers to work out the best way to do them.

**Comment by PeterKietzmann** [comment]
@LudwigKnuepfer would you rebase this PR once more? If not @immesys can you take it over (closing this one and opening an updated PR)

**Comment by immesys** [comment]
I would not mind rebasing, but it didn't actually have any changes, I merge it as-is with no conflicts. In fact at the moment the idle thread (in kernel_init.c) doesn't actually put anything to sleep so merging this file shouldn't even change the behavior of existing code (which is either a feature or a problem depending on how you look at things)

**Comment by PeterKietzmann** [comment]
@immesys I do agree with you but regardless of that I'd like to see Murdock pass, so the PR should be updated. (It's still based on Travis, so kind of old :-) )

**Comment by LudwigKnuepfer** [comment]
I will rebase.

**Comment by LudwigKnuepfer** [comment]
Should I also squash and mark as ready for CI?

**Comment by PeterKietzmann** [comment]
Yes please! But probably I'm just back on Monday.

**Comment by OlegHahm** [comment]
@LudwigKnuepfer, squash?

**Comment by LudwigKnuepfer** [comment]
@OlegHahm I did.

**Comment by biboc** [comment]
Congrats for this PR!

**Pull request merged**